### PR TITLE
[fonts] ability to capitalize the first letter of each word

### PR DESCRIPTION
**Pull request by mkortstiege**
This adds the ability to capitalize the first letter of each word.
Usage: Same as upper-/lowercase by using `[CAPIZALIZE]` or `capitalize` in the font.xml style tag.

Request at forums: http://forum.kodi.tv/showthread.php?tid=220249

### CODE REVIEW

**Comment by tamland** [comment]
I think this style is normally called sentence case.

**Comment by mkortstiege** [comment]
Can't follow. You're just commenting on the wording in the commit and PR message or do you want me to change something?

**Comment by tamland** [comment]
Yeah, call the tag `[SENTENCECASE]` and `ToSentence` or `ToSentenceCase` etc. so it's in line with the other (upper and lower) letter casing.

File: xbmc/utils/StringUtils.cpp [hunk 1]
```diff
@@ -354,6 +355,18 @@ wchar_t toupperUnicode(const wchar_t& c)
   return c;
 }
 
+template <typename Type = char> struct toCapitalize
+{
+   toCapitalize() : bFirst(true) {}
+
+   Type operator()(Type c)
+   {
+      return std::isspace(c) ? (bFirst = true, c) : bFirst ? (bFirst = false, std::toupper(c)) : c;
```
Comment (tamland) [inline xbmc/utils/StringUtils.cpp]: I don't understand any of that:) Can you make it more verbose so it's clearer that is changes state and what else doing on?

**Comment by mkortstiege** [comment]
I am actually reading sentence casing the first time today. For me this is known as capitalizing .. Would like to hear other opinions before changing.

**Comment by mkortstiege** [comment]
An alternative to the templating would be something like this.
```
vector<string> words = Split(str, " ");
for (unsigned int i = 0; i < words.size(); i++)
  words[i][0] = ::toupper(words[i][0]);
str = Join(words, " ");

**Comment by tamland** [comment]
It was more about having two levels of ternaries *with* assignments in them, but on second thought: passing non-pure function to `transform`, probably not a good idea. Transform will not necessary apply that function in-order.

And I think I completely misunderstood which kind of letter casing this was, sorry about that. It' the one listed as Title Case [here](http://en.wikipedia.org/wiki/Letter_case#Stylistic_or_specialised_usage) right? *Not* 'sentence case'.

**Comment by ronie** [comment]
have you tested it with accented characters as well (since we had issues with those in the past)?

and now the bonus question, does it work corectly with RTL languages too?

**Comment by mkortstiege** [comment]
@tamland, this one is Start case. Every first letter is capitalized regardless of function. Objections to change it to the Split/Join method?

@ronie, will test and report back.

File: xbmc/utils/StringUtils.cpp (hunk 1 above)
Comment (Paxxi) [inline xbmc/utils/StringUtils.cpp]: this doesn't compile, std::isspace and std::toupper takes two parameters, you probably meant ::isspace and ::toupper to access the c version of them

**Comment by Paxxi** [comment]
this doesn't work on Windows first because of the std:: I mentioned above and if that's changed to the c-versions it fails because it's not expecting utf8 strings and throws asserts. We also use the C locale everywhere and that is not suitable for these types of string conversions, it'll likely fail on anything except english.

You could make sure it's operating on wstrings and use the uppercase table that's available in stringutils or grab a locale according to user preferences and pass that to isspace and toupper to let it work with the correct casing rules.

**Comment by mkortstiege** [comment]
Updated.

There's still an issues with uppercasing accented characters on FreeBSD, OSX (darwin) and uClibc-based platforms as they don't support `std::toupper` using `std::locale` due to [LangInfo.cpp#L175](https://github.com/xbmc/xbmc/blob/master/xbmc/LangInfo.cpp#L175).

@Fneufneu, @Memphiz mind checking whether or not this limitation is still valid?

**Comment by Memphiz** [comment]
i delegate to @Karlson2k because i think he knows much more about this stuff

**Comment by da-anda** [comment]
I agree with @tamland that capitalize is a bit misleading word for it - reading it at least I would expect this function to either make the string all caps (= uppercase) or small caps ("Kapitälchen" in German).  In PHP the function uppercasing every first letter of a word is called "ucword" (so uppercase word) - not the best either, but [UPPERASE_WORD] would be more selfexplaining. In same way we could add [UPPERCASE_FIRST] to make the first letter of a string caps.
I have no hard feelings about it and am also fine with keeping CAPITALIZE - only prefer blue for the bikeshed :)

**Comment by tamland** [comment]
Your 'uppercase first' is 'sentence case', which is what I was taking about when I said that:P Actually, python uses 'capitalize' too so just pick one

**Comment by Paxxi** [comment]
as @tamland said just pick a name that feels somewhat ok :)
I vote to get this merged as is, it doesn't work for some cases but it's not any worse than our current toupper methods and we can get it sorted out later on when/if we get around to fixing stringutils to handle real unicode.

I would like to see one or more test cases added though, at least one with spaces in the test string :)

@ronie it will work ok for RTL text that's in it's logical representation and represented as LTR with PDF markings to tell the rendererer to flip the string. It will not work if it's stored as RTL and uppercase the last character instead :)

**Comment by mkortstiege** [comment]
Added an additional test string with spaces.

**Comment by Karlson2k** [comment]
In text editor this style is usually named "Titlecase" as it commonly used in titles. I think that name is self-explainable.

**Comment by Karlson2k** [comment]
For the code.
Firstly I'd like to stop producing functions for modifying text in-place. But currently this isn't inline with other transforming functions, so this is optional.
Secondly, locale is created for each letter.
I'd write this function simpler and straightforward:
```` c++
void StringUtils::ToTitlecase(std::wstring &str)
{
  const std::locale& loc = g_langInfo.GetLocale();
  bool isFirstLetter = true;
  for (std::wstring::iterator it = str.begin(); it < str.end(); ++it)
  {
    if (std::isspace(*it, loc))
      isFirstLetter = true;
    else if (isFirstLetter)
    {
      *it = std::toupper(*it, loc);
      isFirstLetter = false;
    }
  }
}
````

**Comment by mkortstiege** [comment]
Thanks for you input. Will update this one tomorrow.

**Comment by Karlson2k** [comment]
The rest looks fine. RTL doesn't matter because text stored in logical order and all RTL alphabets don't have uppercase letters.
Also keep in mind that some languages use combination of letters (digraphs and trigraphs), like 'ch' in Polish and Czech. Luckily for titlecase that doesn't matter. :)

**Comment by mkortstiege** [comment]
Updated. Thanks again for the input.

**Comment by mkortstiege** [comment]
jenkins build this please.

**Comment by Karlson2k** [comment]
I missed test commit!
For complete testing I'd add tests with words with rAndOM case, several spaces before words, spaces at the start of the string, spaces at the end and add tabs and other delimiters.
After run-testing it's ready for merge. :)

**Comment by topfs2** [comment]
In css the term is correct at least, not sure how its done elsewhere.
http://www.w3schools.com/cssref/pr_text_text-transform.asp

**Comment by topfs2** [comment]
A few concerns, to upper isn't always good on all languages AFAIK, but we are using it for uppercase text already I guess so perhaps its out of scope for this PR.

Another one I think worth trying is sentences with double spaces, so double spacing is preserved. Also break lines should also trigger the casing I guess, but perhaps thats handles higher up in the text transform?

**Comment by mkortstiege** [comment]
@topfs2, I think we don't have to worry about languages. It's the skinners decision to use it and to make it an optional skin setting if. We're just offering the ability.

Will check double spaces and line breaks while doing the unit tests.

**Comment by tamland** [comment]
>In text editor this style is usually named "Titlecase" as it commonly used in titles. I think that name is self-explainable.

Sure, but that's not what this is though. Titlecase ignores articles etc. This and 'capitalize' does not.

**Comment by mkortstiege** [comment]
I'll allow the paint to dry and continue this one whenever you guys are ready ;)

**Comment by mkortstiege** [comment]
@Karlson2k added some weird tests. Not sure if it's that what you had in mind..

**Comment by Karlson2k** [comment]
@tamland, "titlecase" means every word starts with Uppercase and rest is lowercase. "Capitalize" looks like all-uppercase. For me "titlecase" is selfexplainable, "capitalize" need additional explanation.
BTW, @mkorstiege if "uppercase" and "lowercase" converts all letters, don't we need to convert rest of the letters to lowercase to be inline with other functions?

**Comment by mkortstiege** [comment]
@Karlson2k that was my first approach as well. Lower all and then upper first but as @HitcherUK pointed out on forums, it looks a bit meh for some like McDonalds -> Mcdonalds or NYPD becomes Nypd.

**Comment by da-anda** [comment]
yes, it should only uppercase first letter and leave the rest unchanged. Btw - what's treated as word boundary? Only spaces so far, right? (ini case I read the code correctly). In any case we shouldn't uppercase string parts connected by a dash. So "Blu-ray" should stay as is and not become Blu-Ray. Maybe also add this as testcase?

Hope I didn't put too much paint on so that it's drying fast ;)

**Comment by mkortstiege** [comment]
Blu-Ray won't happen (dash != space).

**Comment by topfs2** [comment]
Instead of trying to debate what you personally think is the best name, perhaps check what other tools name it. As that has a far larger likelihood of bring universal for more people. We don't always have to recreate the wheel :) as said in css its called capitalise. What is it in other GUI toolkits?

**Comment by topfs2** [comment]
Css also only transform first letter, I.e. leaving McDonalds and NYPD

**Comment by tamland** [comment]
You can chain these functions right? Then only transforming first makes sense as it lets you achieve both.
@Karlson2k Just for reference what would you call the common style (in English at least) of Starting Major Words in Titles and Headers with Uppercase?

**Comment by mkortstiege** [comment]
Squashed and rebased. Will change it to whatever you guys think is the right name.

For the record: I started with Capitalize and still think it's correct.

**Comment by Montellese** [comment]
I vote for Capitalize because that's what CSS uses.

**Comment by topfs2** [comment]
Another concern is that in some languages you uppercase more than a single
letter http://en.m.wikipedia.org/wiki/Capitalization

This is something we would need language knowledge about, and that's not
always the easiest to know. Description tags which are scraped etc. Anyways
I thought it would be interesting to mention. IMO that is something we can
improve in a later PR and shouldn't block this PR.

**Comment by Karlson2k** [comment]
Yep, use "Capitalize". Let's us it like in CCS.

**Comment by mkortstiege** [comment]
Re-renamed to capitalize and rebased. jenkins build this please.

**Comment by tamland** [comment]
Up next: capitalize or capitalise?:D

**Comment by mkortstiege** [comment]
Z please but i have no strong opinion on this. Thanks for annoying the shit out of me ;)

**Comment by da-anda** [comment]
lol. I personally don't care, but our default language is british English, so in case we really want do argue about z vs s I'd say use british spelling :)

**Comment by HitcherUK** [comment]
> Up next: capitalize or capitalise?:D

The skinning engine seems to use American English (eg color, localize, etc) so I guess capitalize.

**Comment by Karlson2k** [comment]
Let's name it exactly the same like in CSS.

**Comment by mkortstiege** [comment]
jenkins build this please

**Comment by topfs2** [comment]
Android
http://developer.android.com/reference/android/widget/TextView.html#attr_android:capitalize

**Comment by mkortstiege** [comment]
jenkins build this please

**Comment by mkortstiege** [comment]
What´s the status here?

**Comment by mkortstiege** [comment]
Will shove this one in tomorrow if there are no objections.

**Comment by HitcherUK** [comment]
@mkortstiege the style tag for fonts this isn't working.

`<style>capitalize</style>`

**Comment by mkortstiege** [comment]
Thanks, will check.

**Comment by mkortstiege** [comment]
https://github.com/xbmc/xbmc/pull/6760 will take care of the style tag issue.

**Comment by HitcherUK** [comment]
@mkortstiege Confirmed this works at the same time as the hint text, many thanks.

**Comment by HitcherUK** [comment]
@mkortstiege Just tried to use this to capitalize a string (id=21429 / string=Order by) but instead of getting 'Order By' I just got '21429'. Tried to add LOCALIZE but that just gave me '[LOCALIZE]21429[/LOCALIZE]' instead.

Thanks.

**Comment by mkortstiege** [comment]
Please post dummy code on the forums so i can easily reproduce. Thanks.

**Comment by HitcherUK** [comment]
Will do.

**Pull request merged**